ratis-server/src/test/java/org/apache/ratis/statemachine/SnapshotManagementTest.java

```diff
@@ -0,0 +1,106 @@
+import org.apache.ratis.server.RaftServer;
+import org.apache.ratis.server.RaftServerConfigKeys;
+import org.apache.ratis.server.impl.MiniRaftCluster;
+import org.apache.ratis.server.impl.RaftServerTestUtil;
+import org.apache.ratis.server.raftlog.RaftLog;
+import org.apache.ratis.server.simulation.MiniRaftClusterWithSimulatedRpc;
+import org.apache.ratis.util.JavaUtils;
+import org.apache.ratis.util.Log4jUtils;
+import org.junit.Assert;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+public abstract class SnapshotManagementTest extends BaseTest {
+
+  {
+    Log4jUtils.setLogLevel(RaftServer.Division.LOG, Level.DEBUG);
+    Log4jUtils.setLogLevel(RaftLog.LOG, Level.INFO);
+    Log4jUtils.setLogLevel(RaftClient.LOG, Level.INFO);
+  }
+
+  static final Logger LOG = LoggerFactory.getLogger(SnapshotManagementTest.class);
+  private MiniRaftCluster cluster;
+
+  public MiniRaftCluster.Factory<?> getFactory() {
+    return MiniRaftClusterWithSimulatedRpc.FACTORY;
+  }
+
+  @Test
+  public void testTakeSnapshot() throws Exception {
+    final RaftProperties prop = new RaftProperties();
+    prop.setClass(MiniRaftCluster.STATEMACHINE_CLASS_KEY,
+          SimpleStateMachine4Testing.class, StateMachine.class);
+    RaftServerConfigKeys.Snapshot.setAutoTriggerEnabled(prop, false);
+    cluster = getFactory().newCluster(1,prop);
+    cluster.start();
+
+    int i = 0;
+    try {
+      final RaftServer.Division leader = RaftTestUtil.waitForLeader(cluster);
+      final RaftPeerId leaderId = leader.getId();
+      try(final RaftClient client = cluster.createClient(leaderId)) {
+        //todo(yaolong liu) : make 5 to be a configurable value
+        for (; i < 5; i++) {
+          RaftClientReply reply = client.io().send(new RaftTestUtil.SimpleMessage("m" + i));
+          Assert.assertTrue(reply.isSuccess());
+        }
+        final SnapshotRequest r = new SnapshotRequest(client.getId(), leaderId, cluster.getGroupId(),
+              CallId.getAndIncrement(), 3000);
+        RaftServerTestUtil.takeSnapshotAsync(leader, r);
+      }
+
+      // wait for the snapshot to be done
+      long nextIndex = cluster.getLeader().getRaftLog().getNextIndex();
+      LOG.info("nextIndex = {}", nextIndex);
+
+      final List<File> snapshotFiles = LongStream.range(0, nextIndex)
+            .mapToObj(j ->
+                 SimpleStateMachine4Testing
+                        .get(cluster.getLeader())
+                        .getStateMachineStorage()
+                        .getSnapshotFile(cluster.getLeader().getInfo().getCurrentTerm(), j))
+            .collect(Collectors.toList());
+      JavaUtils.attemptRepeatedly(() -> {
+        Assert.assertTrue(snapshotFiles.stream().anyMatch(File::exists));
+        return null;
+      }, 100, ONE_SECOND, "snapshotFile.exist", LOG);
```

Comment:
Once we have the snapshot index, we can check only the expected file.

Reply:
Good change, unit test is more accurate